eras/conway/impl/src/Cardano/Ledger/Conway/Rules/Gov.hs

```diff
@@ -276,65 +258,102 @@ actionWellFormed ga = failureUnless isWellFormed $ MalformedProposal ga
-        (GovAction era -> Bool) ->
-        Bool
-      isValidPrevGovActionId prevGovActionId proposalsCond =
-        case prevGovActionId of
-          -- The case of having an SNothing as valid, for the very first proposal ever, is handled in `prevActionAsExpected`
-          SNothing -> False
-          SJust (PrevGovActionId govActionId) ->
-            case proposalsLookupId govActionId proposalss of
-              Nothing -> False
-              -- lookup has to succeed _and_ purpose of looked-up action has to match condition
-              Just found -> proposalsCond $ gasAction found
-      isValid proposal =
-        prevActionAsExpected (proposal ^. pProcGovActionL) prevGovActionIds
-          || case proposal ^. pProcGovActionL of
-            ParameterChange prev _ ->
-              isValidPrevGovActionId prev $ \case ParameterChange {} -> True; _ -> False
-            HardForkInitiation prev _ ->
-              isValidPrevGovActionId prev $ \case HardForkInitiation {} -> True; _ -> False
-            TreasuryWithdrawals _ -> True
-            NoConfidence prev ->
-              isValidPrevGovActionId prev $ \case NoConfidence {} -> True; UpdateCommittee {} -> True; _ -> False
-            UpdateCommittee prev _ _ _ ->
-              isValidPrevGovActionId prev $ \case NoConfidence {} -> True; UpdateCommittee {} -> True; _ -> False
-            NewConstitution prev _ ->
-              isValidPrevGovActionId prev $ \case NewConstitution {} -> True; _ -> False
-            InfoAction -> True
-      invalidProposals = Seq.filter (not . isValid) $ SSeq.fromStrict $ OSet.toStrictSeq $ procedures ^. govProceduresProposalsL
-   in failureUnless (Seq.null invalidProposals) $ InvalidPrevGovActionIdsInProposals invalidProposals
+  PrevGovActionIdsChildren era ->
+  Maybe (PrevGovActionIdsChildren era)
+addToPrevGovActionIdsChildren actionId action PrevGovActionIds {..} children =
+  case action of
+    ParameterChange prev _ ->
+      go pgaPParamUpdate prev pgacPParamUpdateL actionId children
+    HardForkInitiation prev _ ->
+      go pgaHardFork prev pgacHardForkL actionId children
+    TreasuryWithdrawals _ -> Just children
+    NoConfidence prev ->
+      go pgaCommittee prev pgacCommitteeL actionId children
+    UpdateCommittee prev _ _ _ ->
+      go pgaCommittee prev pgacCommitteeL actionId children
+    NewConstitution prev _ ->
+      go pgaConstitution prev pgacConstitutionL actionId children
+    InfoAction -> Just children
+  where
+    go ::
+      StrictMaybe (PrevGovActionId purpose (EraCrypto era)) ->
+      StrictMaybe (PrevGovActionId purpose (EraCrypto era)) ->
+      Lens' (PrevGovActionIdsChildren era) (Set.Set (GovActionId (EraCrypto era))) ->
+      GovActionId (EraCrypto era) ->
+      PrevGovActionIdsChildren era ->
+      Maybe (PrevGovActionIdsChildren era)
+    go pgai prev lens' actionId' children' =
+      if pgai == prev
+        then Just $ children' & lens' %~ Set.insert actionId'
+        else Nothing
```

Comment:
The type signature for `go` takes more space then the function actually saves. Also it forces you to use this lens that bypasses the designed type safety: 
```
Lens' (PrevGovActionIdsChildren era) (Set.Set (GovActionId (EraCrypto era)))
```

```suggestion
  case action of
    ParameterChange prev _ -> do
      guard (pgaPParamUpdate == prev)
      let prevGovActionId :: PrevGovActionId 'PParamUpdatePurpose c
          prevGovActionId = PrevGovActionId actionId
      Just $ children & pgacPParamUpdateL %~ insertPrevActionId prevGovActionId
    HardForkInitiation prev _ -> do
      guard (pgaHardFork == prev)
      let prevGovActionId :: PrevGovActionId 'HardForkPurpose c
          prevGovActionId = PrevGovActionId actionId
      Just $ children & pgacHardForkL %~ insertPrevActionId prevGovActionId
    TreasuryWithdrawals _ -> Just children
    NoConfidence prev -> do
      guard (pgaCommittee == prev)
      let prevGovActionId :: PrevGovActionId 'CommitteePurpose c
          prevGovActionId = PrevGovActionId actionId
      Just $ children & pgacCommitteeL %~ insertPrevActionId prevGovActionId
    UpdateCommittee prev _ _ _ -> do
      guard (pgaCommittee == prev)
      let prevGovActionId :: PrevGovActionId 'CommitteePurpose c
          prevGovActionId = PrevGovActionId actionId
      Just $ children & pgacCommitteeL %~ insertPrevActionId prevGovActionId
    NewConstitution prev _ -> do
      guard (pgaConstitution == prev)
      let prevGovActionId :: PrevGovActionId 'ConstitutionPurpose c
          prevGovActionId = PrevGovActionId actionId
      Just $ children & pgacConstitutionL %~ insertPrevActionId prevGovActionId
    InfoAction -> Just children
  where
    insertPrevActionId :: 
      PrevGovActionChildren r c -> PrevGovActionId r c -> PrevGovActionChildren r c 
    insertPrevActionId (PrevGovActionChildren m) prevGovActionId = 
      PrevGovActionChildren (Set.insert prevGovActionId m)
```

Reply:
After implementing the newtype wrapper preserving type-safely, this turned out to be needing very little change. We don't need the type signature anymore. And the implementation is shorter with the use of `go`. But I will be happy to switch to your suggested way of implementing this if you still prefer it, let me know :heart: